### PR TITLE
Polyline

### PR DESCRIPTION
**Pull request by patriciogonzalezvivo**
First round

### CODE REVIEW

**Comment by bcamper** [comment]
Note: we lost some functionality in this change that needs to be restored before it can be merged. Namely:

* Handling of closed polygons outlines, which form a continuous loop (instead of having caps on the first and last points). See `closed_polygon` flag.
* Automatic removal of line segments falling on tile edges, which is needed for correctly rendering polygon outlines, since many logical polygons have been split into several polygons across tile boundaries (we are basically visually reassembling it to appear as a single polygon). See `remove_tile_edges` flag.

File: src/gl/gl_builders.js [hunk 1]
```diff
@@ -145,213 +145,178 @@ GLBuilders.buildPolylines = function (
...
+        if (lineSize < 2) {
+            continue;
+        }
 
-        vertices.push(
-            pb_inner, pb_outer, pa_inner,
-            pa_inner, pb_outer, pa_outer
-        );
+        //  Initialize variables
+        var normPrevCurr = [0.0,0.0,0.0],  // Right normal to segment between previous and current m_points
```
Comment (bcamper) [inline src/gl/gl_builders.js]: 0.0 not needed here, can just be 0.

File: src/gl/gl_builders.js [hunk 2]
```diff
@@ -145,213 +145,178 @@ GLBuilders.buildPolylines = function (
...
+            normCurrNext = [0.0,0.0,0.0],  // Right normal to segment between current and next m_points
+            rightNorm = [0.0,0.0,0.0];     // Right "normal" at current point, scaled for miter joint
+
+        var prevCoord = [], // Previous point coordinates
+            currCoord = [], // Current point coordinates
+            nextCoord = []; // Next point coordinates
+
+        //  Compute first pair of vertexes based on the second point of the line
+        currCoord = Vector.set( line[0] );
```
Comment (bcamper) [inline src/gl/gl_builders.js]: For style consistency, please remove extra spaces in parens.

File: src/gl/gl_builders.js [hunk 3]
```diff
@@ -145,213 +145,178 @@ GLBuilders.buildPolylines = function (
...
+        //  b) add the pair vertex into the vertex buffer
+        if (scaling_index) {
+            //  b.a If scaling is on add the vertex (the currCoord) and the scaling Vecs (normals pointing where to extrude the vertexes)
+            vertices.push(currCoord);
+            vertices.push(currCoord);
+            scalingVecs.push( rightNorm );
+            scalingVecs.push( Vector.neg(rightNorm) );
+        } else {
+            //  b.b Add the extruded vertexes
+            vertices.push([ currCoord[0] + rightNorm[0] * halfWidth,
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Ditto extra spaces in parens.

File: src/gl/gl_builders.js [hunk 4]
```diff
@@ -145,213 +145,178 @@ GLBuilders.buildPolylines = function (
...
+        }
 
-        // Add UVs
+        // c) Add uv's if they are enable
         if (texcoord_index) {
-            texcoords.push(
-                [min_u, min_v], [max_u, min_v], [min_u, max_v],
-                [min_u, max_v], [max_u, min_v], [max_u, max_v]
-            );
+            texcoords.push( [max_u, min_v], [min_u, min_v] );
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Is this right? 3 references to min but only 1 to max? (I might just not see what's happening here.)

File: src/gl/gl_builders.js [hunk 5]
```diff
@@ -145,213 +145,178 @@ GLBuilders.buildPolylines = function (
...
+            addVertex(2*i+2);
+            addVertex(2*i+1);
+            addVertex(2*i+0);
+
+            addVertex(2*i+2);
+            addVertex(2*i+3);
+            addVertex(2*i+1);
         }
 
+        // // Add triangles indices
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Can remove this commented code.

File: src/gl/gl_builders.js [hunk 6]
```diff
@@ -145,213 +145,178 @@ GLBuilders.buildPolylines = function (
...
+        //     indices_data.push(vertexData_Offset + 2*i+2);
+        //     indices_data.push(vertexData_Offset + 2*i+1);
+        //     indices_data.push(vertexData_Offset + 2*i);
+        //     indices_data.push(vertexData_Offset + 2*i+2);
+        //     indices_data.push(vertexData_Offset + 2*i+3);
+        //     indices_data.push(vertexData_Offset + 2*i+1);
+        // }
+
+        // Clean all
+        vertices.length = 0;
```
Comment (bcamper) [inline src/gl/gl_builders.js]: See other comments, please use `= []` instead (and move to top of this loop).

File: src/vector.js [hunk 7]
```diff
@@ -2,6 +2,107 @@
...
+        // Divide to vectors
+        var len = Math.min(v1.length,v2.length);
+        for (i = 0; i < len; i++) {
+            v[i] = v1[i] / v2[i];
+        }
+    }
+    return v;
+};
+ 
+// Get 2D perpendicular FAST
```
Comment (bcamper) [inline src/vector.js]: 1. Is this used anywhere?
2. Why the FAST comment?
3. It looks like it won't work, since the arguments are vectors and you're doing subtraction on the objects directly (remember, this isn't C++ :)

File: src/vector.js [hunk 8]
```diff
@@ -2,6 +2,107 @@
...
+    return [ v2[1] - v1[1], 
+             v1[0] - v2[0] ];
+};
+
+// Get 2D vector rotated 
+Vector.rot = function (v, a) {
+    var V = v;
+    var vr = Vector.length(v);
+    var va = Vector.angle(v);
+    V[0] = vr * Math.cos(va+a);
```
Comment (bcamper) [inline src/vector.js]: Ditto as above, these are JS objects, you don't have any overloaded operators.

File: src/vector.js [hunk 9]
```diff
@@ -2,6 +2,107 @@
...
+};
+ 
+// Get 2D perpendicular FAST
+Vector.perp = function (v1, v2) {
+    return [ v2[1] - v1[1], 
+             v1[0] - v2[0] ];
+};
+
+// Get 2D vector rotated 
+Vector.rot = function (v, a) {
```
Comment (bcamper) [inline src/vector.js]: Objects are passed by reference, this is going to overwrite the contents of the argument vector `v`. Set `V = []` to make a new vector instead.

File: src/vector.js [hunk 10]
```diff
@@ -53,6 +154,17 @@ Vector.cross  = function (v1, v2)
     ];
 };
 
+// Dot product of two vectors
+Vector.dot = function (v1, v2)
+{
+    var n = 0.0;
```
Comment (bcamper) [inline src/vector.js]: 0.0 is superfluous (very nitpicky I know, just FYI).

File: src/vector.js [hunk 11]
```diff
@@ -2,6 +2,107 @@
...
+    var V = v;
+    var vr = Vector.length(v);
+    var va = Vector.angle(v);
+    V[0] = vr * Math.cos(va+a);
+    V[1] = vr * Math.sin(va+a);
+    return V;
+};
+
+// Get 2D heading angle
+Vector.angle = function (v) {
```
Comment (iwillig) [inline src/vector.js]: Destructing allows you to write this function like this in more compact manner,
```javascript
Vector.angle = function ([x, y]) {
    return Math.atan2(x, y);
};
```
There should be no overhead to form.

File: src/vector.js [hunk 12]
```diff
@@ -53,6 +154,17 @@ Vector.cross  = function (v1, v2)
     ];
 };
 
+// Dot product of two vectors
+Vector.dot = function (v1, v2)
+{
+    var n = 0.0;
+    var lim = Math.min(v1.length, v2.length);
```
Comment (bcamper) [inline src/vector.js]: While in some ways better to check min length of each argument (avoids any invalid accesses), it's also kind of weird, because it's not really valid to call this with arguments of mismatching length anyway, is it? For functions that might be called frequently in inner loops, we would often not carefully check arguments, but instead assume that the calling code is going to call it correctly, and just let the function throw an error if there is an invalid access. So in that case you could just use the length of the first argument and assume they are the same. I think the same may apply to some other functions above where you're checking the min of multiple arguments too?

Comment (bcamper) [inline src/vector.js]: @iwillig your opinion here?

File: src/vector.js (hunk 11 above)
Comment (bcamper) [inline src/vector.js]: Good point - @patriciogonzalezvivo destructuring is a new feature of ES6 (you may have seen it in python): https://github.com/lukehoban/es6features#destructuring

File: src/vector.js (hunk 12 above)
Comment (iwillig) [inline src/vector.js]: I agree with Brett, it seems fine to just use the length of the first argument. One thing to mention is that JavaScript Arrays do not throw out of bound errors for element they don't have. They return undefined, so 

```javascript
var a = [1,2,3,4];
a[10] === undefined; // true
```
So if you do have a shorter Array as the first argument, the multiplication operation will produce a 'NaN'.

Comment (iwillig) [inline src/vector.js]: It actually gets better, because testing to see if you have a NaN value is hard. In JS
```javascript
NaN === NaN; // false
NaN == NaN; // false
```
The correct way to detect if you have a NaN value is to use Object.is, now. I think this is new.
```javascript
Object.is(NaN, NaN); // true
```

File: src/gl/gl_builders.js [hunk 13]
```diff
@@ -145,213 +145,149 @@ GLBuilders.buildPolylines = function (
...
-                    else if (!edge1) {
-                        buildSegment(anchors[p][0], anchors[p][1]);
-                    }
-                    else if (!edge2) {
-                        buildSegment(anchors[p][1], anchors[p][2]);
-                    }
-                }
-            }
+    // Add to equidistant pairs of vertices
+    function addVertexPair ( coord, normal, v_pct ) {
```
Comment (iwillig) [inline src/gl/gl_builders.js]: Spacing, normally we don't put spaces around the function arguments.

File: src/gl/gl_builders.js [hunk 14]
```diff
@@ -145,213 +145,149 @@ GLBuilders.buildPolylines = function (
...
-            if (len_sq > (width * width * miter_len_max * miter_len_max)) {
-                line_debug = 'distance';
-                intersect_outer = Vector.normalize([intersect_outer[0] - joint[0], intersect_outer[1] - joint[1]]);
-                intersect_outer = [
-                    joint[0] + intersect_outer[0] * miter_len_max,
-                    joint[1] + intersect_outer[1] * miter_len_max
-                ];
-            }
+        // Do this with the rest (except the last one)
+        for(i = 1; i < lineSize - 1 ; i++){
```
Comment (iwillig) [inline src/gl/gl_builders.js]: Missing a space here.

File: src/gl/gl_builders.js [hunk 15]
```diff
@@ -145,213 +145,149 @@ GLBuilders.buildPolylines = function (
...
+                            [ min_u, (1-v_pct)*min_v + v_pct*max_v ]);
         }
     }
 
-    // Add vertices to buffer
-    for (var v=0; v < vertices.length; v++) {
-        vertex_template[0] = vertices[v][0];
-        vertex_template[1] = vertices[v][1];
+    // Add a vertex based on the index position into the VBO 
+    function addIndex ( index ) {
```
Comment (iwillig) [inline src/gl/gl_builders.js]: Not sure if this is important, but defining a function within another function means that on every function invocation we are going to be creating a new function object. 

Normally this is not a huge deal in JS, but you mentioned that some of this code is in a tight inner loop.

File: src/gl/gl_builders.js [hunk 16]
```diff
@@ -145,213 +145,149 @@ GLBuilders.buildPolylines = function (
...
-
-                pb_inner[1], pb_outer[1], pb_inner[0],
-                pb_inner[0], pb_outer[1], pb_outer[0]
-            );
+
+        rightNorm = Vector.normalize( normCurrNext );
+        addVertexPair(nextCoord, rightNorm, 1 );
+
+        // Add vertices to buffer acording their index
+        for (i = 0; i < lineSize - 1; i++) {
```
Comment (iwillig) [inline src/gl/gl_builders.js]: I noticed that you reuse the `i` variable in this function. Feel free to use `let` instead of `var`. `Let` respects block scoping while `var` only understands function scope. 

https://github.com/google/traceur-compiler/wiki/LanguageFeatures#block-scoped-binding
I believe this is turned on by default now in traceur

Comment (patriciogonzalezvivo) [inline src/gl/gl_builders.js]: great tip! thanks

Comment (bcamper) [inline src/gl/gl_builders.js]: Except I think that traceur emulates this with try/catch blocks, and they warn it is slow? :/  I've used it for non-performance-critical code, but have left it out of frequently called functions.

Comment (iwillig) [inline src/gl/gl_builders.js]: Yeah I agree that we should not use features that could cause tangram to be slow. I looked into how they implemented `let` in traceur. For the following code, 
```javascript
let a = [1, 2, 3, 4];
let i = 10;
for (let i in a) {
    console.log(i);
}

function do_something() {
  console.log(foo); // ReferenceError
  let foo = 2;
}

if (true) {
    let x = 'a';
    console.log(x);
    if (true) {
        let x = 'b';
        console.log(x);
        if (true) {
            let x = 'c';
            console.log(x);
            if (true) {
                let x = 'd';
                console.log(x);
            }
        }
    }
}
```
traceur generates 
```javascript
System.registerModule("test.js", [], function() {
  "use strict";
  var __moduleName = "test.js";
  var a = [1, 2, 3, 4];
  var i = 10;
  for (var i$__0 in a) {
    console.log(i$__0);
  }
  function do_something() {
    console.log(foo);
    var foo = 2;
  }
  if (true) {
    var x = 'a';
    console.log(x);
    if (true) {
      var x$__1 = 'b';
      console.log(x$__1);
      if (true) {
        var x$__2 = 'c';
        console.log(x$__2);
        if (true) {
          var x$__3 = 'd';
          console.log(x$__3);
        }
      }
    }
  }
  return {};
});
System.get("test.js" + '');
```
I need to make sure that traceur always uses this method for compiling `let`, but if so then it should be no or minimal overhead.
 
I can look into this more tomorrow as there might be situations where they need try and catch.

Comment (bcamper) [inline src/gl/gl_builders.js]: Oh nice, that's great. I thought I read the catch/try thing on the traceur docs, but that was also before they enabled it by default. Perhaps they switched to this method.

File: demos/styles.yaml [hunk 17]
```diff
@@ -257,7 +257,7 @@ layers:
             filter: water
 
         style:
-            # name: water
+            name: water
```
Comment (bcamper) [inline demos/styles.yaml]: Not sure why this change ended up here but could you revert? I turned water animation off by default in the main demo.

File: src/gl/gl_builders.js [hunk 18]
```diff
@@ -134,9 +134,59 @@ GLBuilders.buildExtrudedPolygons = function (
...
+
+    // c) Add uv's if they are enable
+    if (texcoords) {
+        texcoords.push( [ max_u, (1-v_pct)*min_v + v_pct*max_v ],
+                        [ min_u, (1-v_pct)*min_v + v_pct*max_v ]);
+    }
+}
+
+// Add a vertex based on the index position into the VBO 
+    function addIndex (index, { vertex_data, vertex_template, halfWidth, vertices, scaling_index, scalingVecs, texcoord_index, texcoords }) {
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Remove extra indentation.

File: src/gl/gl_builders.js [hunk 19]
```diff
@@ -134,9 +134,59 @@ GLBuilders.buildExtrudedPolygons = function (
...
+        //  Add vertex to VBO
+        vertex_data.addVertex(vertex_template);
+    }
+
 // Build tessellated triangles for a polyline
-// Basically following the method described here for miter joints:
-// http://artgrammer.blogspot.co.uk/2011/07/drawing-polylines-by-tessellation.html
 GLBuilders.buildPolylines = function (
     lines,
     z, width,
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Remove `z` argument (not used anymore).

File: src/gl/gl_builders.js [hunk 20]
```diff
@@ -145,213 +195,99 @@ GLBuilders.buildPolylines = function (
...
-
-                pb_inner[1], pb_outer[1], pb_inner[0],
-                pb_inner[0], pb_outer[1], pb_outer[0]
-            );
+            addIndex(2*i+2, constants);
+            addIndex(2*i+3, constants);
+            addIndex(2*i+1, constants);
         }
 
+        // for (var i = 0; i < lineSize - 1; i++) {
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Remove commented code.

File: src/gl/gl_builders.js [hunk 21]
```diff
@@ -134,9 +134,59 @@ GLBuilders.buildExtrudedPolygons = function (
     }
 };
 
+// Add to equidistant pairs of vertices
+function addVertexPair (coord, normal, v_pct, { halfWidth, vertices, scalingVecs, texcoords, min_u, min_v, max_u, max_v }) {
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Maybe add a comment for these functions indicating that they are internal helper methods for the polyline builder. Maybe put them after the main polyline builder method.

File: src/gl/gl_builders.js [hunk 22]
```diff
@@ -145,213 +195,99 @@ GLBuilders.buildPolylines = function (
...
-        pb,
+        scalingVecs = scaling_index && [],   //  vertices directions for then (on GLSL vertex shader) extrude
+        texcoords = texcoord_index && [],
+        halfWidth = width/2,
         num_lines = lines.length;
-
+        
     var [[min_u, min_v], [max_u, max_v]] = texcoord_scale || [[0, 0], [1, 1]];
 
+    var constants = { vertex_data, vertex_template, halfWidth, vertices, scaling_index, scalingVecs, texcoord_index, texcoords, min_u, min_v, max_u, max_v };
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Maybe add a comment like:
`// Values that are constant for each line and are passed to helper functions`

File: src/gl/gl_builders.js [hunk 23]
```diff
@@ -145,213 +195,99 @@ GLBuilders.buildPolylines = function (
...
     }) {
 
-    // Build triangles
+    // Build variables
     var vertices = [],
-        texcoords = [],
-        p,
-        pa,
-        pb,
+        scalingVecs = scaling_index && [],   //  vertices directions for then (on GLSL vertex shader) extrude
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Initialize these inside the top of the line loop instead.

File: src/gl/gl_builders.js [hunk 24]
```diff
@@ -145,213 +195,99 @@ GLBuilders.buildPolylines = function (
...
+        // for (var i = 0; i < lineSize - 1; i++) {
+        //     indices_data.push(vertexData_Offset + 2*i+2);
+        //     indices_data.push(vertexData_Offset + 2*i+1);
+        //     indices_data.push(vertexData_Offset + 2*i);
+        //     indices_data.push(vertexData_Offset + 2*i+2);
+        //     indices_data.push(vertexData_Offset + 2*i+3);
+        //     indices_data.push(vertexData_Offset + 2*i+1);
+        // }
+
+        // Clean all
```
Comment (bcamper) [inline src/gl/gl_builders.js]: Re-initialize these to `[]` at the top of the line loop instead.

File: src/gl/gl_builders.js (hunk 23 above)
Comment (patriciogonzalezvivo) [inline src/gl/gl_builders.js]: you mean line and lineSize?

**Comment by patriciogonzalezvivo** [comment]
Polyline ready for review. Not caps and joins implemented yet

**Comment by meetar** [comment]
So does this just silently replace the old line style? Do we have or want a way to specify eg endcap style in the scene def?

**Comment by bcamper** [comment]
Yep this is just a different way of implementing what we had before. There
aren't any additions for caps and joints yet, but there should be soon!
